enhancements/monitoring/replace-prometheus-adapter-with-metrics-server.md

```diff
@@ -0,0 +1,202 @@
+* Allow Prometheus Adapter to Metrics Server transition (and vice-versa) through a configuration or feature-flag.
+
+
+### Non-Goals
+
+* Allow optional prometheus stack
+
+## Proposal
+
+The proposal is to replace the Prometheus Adapter with the Metrics Server for monitoring resource metrics in OpenShift.
+
+### Workflow Description
+
+The workflow for using the Metrics Server in OpenShift for monitoring resource metrics would involve the following steps:
+
+1. User sets `enableMetricsServer: true` in the ConfigMap `openshift-monitoring/cluster-monitoring-config`
+
+    ```yaml
+    apiVersion: v1
+    kind: ConfigMap
+    metadata:
+      name: cluster-monitoring-config
+      namespace: openshift-monitoring
+    data:
+      config.yaml: |
+        enableMetricsServer: true 
+    ```
+
+1. Deploy and configure the metrics-server in the `openshift-monitoring` namespace
+1. Verify `metrics-server` resource API is functional by querying metrics-server endpoint (TBD)
+1. Update `APIService` resource to point to metrics-server API instead of `prometheus-adapter` API
+1. Delete/Scale Down `prometheus-adapter` deployment
+
+#### Variation [optional]
+
+NA
+
+### API Extensions
+
+- Addition of a new field to the `cluster-monitoring-config` ConfigMap
+
+  ```go
+  type ClusterMonitoringConfiguration struct {
+    // Enable MetricsServer
+    EnableMetricsServer bool `json:"enableMetricsServer,omitempty"`
+  }
+  ```
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Configuration/Enablement
+
+Metrics Server is not enabled by default. It must be explicitly enabled by the user. It is enabled by setting 
+`enableMetricsServer: true` in the ConfigMap `openshift-monitoring/cluster-monitoring-config`
+
+**Note:** In the long run, the existing cluster-monitoring-config which is currently a config-map will be moved to a Custom Resource  and is out of scope for this enhancement.
+
+#### Resource Impact
+
+Once metrics-server is enabled, by default 2 replicas of metrics-server instances are being deployed in the openshift-monitoring namespace.
+
+As of OpenShift 4.13 the following idle resource usage have been measured:
+
+Kubelet: TBD
+Metrics Server: TBD
+
+**TODO:** Capture the details on performance testing done from [MON-3210](https://issues.redhat.com/browse/MON-3210)
+
+
+### Risks and Mitigations
+
+One of the risk is resource impact on larger clusters. We need to do performance testing on larger OpenShift cluster to measure the impact on Kubelet when metrics-server is deployed.
+
+[MON-3210](https://issues.redhat.com/browse/MON-3210) is created for doing this.
+
+### Drawbacks
+
+N/A
+
+
+## Design Details
+
+### Open Questions [optional]
+
+[custom-metrics-autoscaler](https://github.com/openshift/custom-metrics-autoscaler-operator) also deploys metrics-server when user enables it, will this affect metrics-server we deploy in monitoring stack?
```

Comment:
Can you link to where it's happening? This is new to me.

Reply:
when keda controller is created after installing operator I see metrics-server is deployed as well
https://github.com/openshift/custom-metrics-autoscaler-operator/blob/39aa2b9d5a0db2206d4963fe2e551178e7d369ad/controllers/keda/kedacontroller_controller.go#L204C2-L210.